tiledb/sm/filesystem/s3.cc

```diff
@@ -255,58 +267,83 @@ Status S3::flush_object(const URI& uri) {
   Aws::Http::URI aws_uri = uri.c_str();
   std::string path_c_str = aws_uri.GetPath().c_str();
 
-  // Take a lock protecting the shared multipart data structures
+  // Take a lock protecting 'multipart_upload_states_'.
   std::unique_lock<std::mutex> multipart_lck(multipart_upload_mtx_);
 
   // Do nothing - empty object
-  auto multipart_upload_it = multipart_upload_.find(path_c_str);
-  if (multipart_upload_it == multipart_upload_.end())
+  auto state_iter = multipart_upload_states_.find(path_c_str);
+  if (state_iter == multipart_upload_states_.end())
     return Status::Ok();
 
-  // Get the completed upload object
-  auto completed_multipart_upload = multipart_upload_it->second;
+  const MultiPartUploadState& state = state_iter->second;
 
-  // Add all the completed parts (sorted by part number) to the upload object.
-  auto completed_parts_it = multipart_upload_completed_parts_.find(path_c_str);
-  if (completed_parts_it == multipart_upload_completed_parts_.end()) {
-    return LOG_STATUS(Status::S3Error(
-        "Unable to find completed parts list for S3 object " +
-        uri.to_string()));
+  if (state.st.ok()) {
+    Aws::S3::Model::CompleteMultipartUploadRequest complete_request =
+        build_multipart_upload_request(state);
+    auto outcome = client_->CompleteMultipartUpload(complete_request);
+
+    return finish_flush_object(
+        std::move(outcome), uri, std::move(multipart_lck), state, buff);
```

Comment:
I'm assuming the mutex move is required here because the semantics of your functions require that the lock continues to be held across the function call without unlocking and relocking. If that's the case, can you add a brief comment here stating that?

Reply:
Right -- the lock is protecting 'state' since it's a const-ref. We don't want to unlock the shared state and allow another thread to mutate+invalidate the reference. I've documented in the header that the lock protects 'state', but I'll add further explanation in the .cc

Reply:
Obsolete